### PR TITLE
fix(smartplaylist): reject NSP mixing top-level 'any' and 'all'

### LINKED ISSUE #5757: [Bug]: Smart Playlists with multiple groups (e.g. any + all) seem to be allowed but are not saved to db by scanner
### I confirm that:

- [x] I have searched the existing [open AND closed issues](https://github.com/navidrome/navidrome/issues?q=is%3Aissue) to see if an issue already exists for the bug I've encountered
- [x] I'm using the latest version (your issue may have been fixed already)

### Version

0.63.0

### Current Behavior

Based on an example on the Smart Playlists documentation page it would seem that a playlist with any + all should be allowed, e.g. this example which has an any + all + sort:
```
{
  "name": "Overplayed Favorites",
  "comment": "Most Played Favorites Played Within Last 4yr",
  "public": true,
  "any": [{ "inPlaylist": { "path": "most-played-favorites.nsp" } }],
  "all": [{ "notInPlaylist": { "path": "favorites-not-played-in-4-yrs.nsp" } }],
  "sort": "playCount, lastPlayed"
}
```

However when I copy this exact example and import it then it does not seem to be imported completely into the db, only the first rule group (any) gets imported and stored:
```
navidrome D select name, path, updated_at, rules from playlist where name = 'Overplayed Favorites';
|         name         |                     path                     |             updated_at              |                                            rules                                             |
|----------------------|----------------------------------------------|-------------------------------------|----------------------------------------------------------------------------------------------|
| Overplayed Favorites | /music/playlists/admin/ExampleOverplayed.nsp | 2026-07-10 20:43:38.145072577+00:00 | {"any":[{"inPlaylist":{"path":"most-played-favorites.nsp"}}],"sort":"playCount, lastPlayed"} |
```

If I take the same example, change the name/filename, flip the rule groups:
```
{
  "name": "Overplayed Favorites Flipped",
  "comment": "Most Played Favorites Played Within Last 4yr, all then any",
  "public": true,
  "all": [{ "notInPlaylist": { "path": "favorites-not-played-in-4-yrs.nsp" } }],
  "any": [{ "inPlaylist": { "path": "most-played-favorites.nsp" } }],
  "sort": "playCount, lastPlayed"
}
```

Then the db shows the same thing as the first one oddly:
```
navidrome D select name, path, updated_at, rules from playlist where name = 'Overplayed Favorites Flipped';
|             name             |                        path                         |             updated_at              |                                            rules                                             |
|------------------------------|-----------------------------------------------------|-------------------------------------|----------------------------------------------------------------------------------------------|
| Overplayed Favorites Flipped | /music/playlists/admin/ExampleOverplayedFlipped.nsp | 2026-07-10 20:43:38.147794813+00:00 | {"any":[{"inPlaylist":{"path":"most-played-favorites.nsp"}}],"sort":"playCount, lastPlayed"} |
```

If I completely remove the any group then it stores the all group (ignore the comment, I forgot to update it):
```
{
  "name": "Overplayed Favorites All Only",
  "comment": "Most Played Favorites Played Within Last 4yr, all then any",
  "public": true,
  "all": [{ "notInPlaylist": { "path": "favorites-not-played-in-4-yrs.nsp" } }],
  "sort": "playCount, lastPlayed"
}
```

```
navidrome D select name, path, updated_at, rules from playlist where name like 'Overplayed Favorites All Only';
|             name              |                        path                         |             updated_at              |                                                  rules                                                  |
|-------------------------------|-----------------------------------------------------|-------------------------------------|---------------------------------------------------------------------------------------------------------|
| Overplayed Favorites All Only | /music/playlists/admin/ExampleOverplayedAllOnly.nsp | 2026-07-10 20:54:08.859511261+00:00 | {"all":[{"notInPlaylist":{"path":"favorites-not-played-in-4-yrs.nsp"}}],"sort":"playCount, lastPlayed"} |
```

Note that I do not have the other playlists (e.g. favorites-not-played-in-4-yrs.nsp) so I dont expect these to actually work/show songs. I arrived at testing like this by way of troubleshooting my own playlists that have any + all rule groups. I tested in Feishin and its Smart Playlist creation UI wont allow two rule groups to be created, only one any or one all. I know this isnt definitive since its not yet a stable API/etc.

Thinking that the scanner is doing something to select only one group but not being proficient at Golang to figure out the code side of it, I decided to just update the db directly.

This is the playlist I am actually aiming for to work:
```
{
  "name": "Radio",
  "comment": "All songs played in last 4 days, 25 random never played, not 1 star, sorted random",
  "any": [
    {"inPlaylist": {"id": "QtfW3w5x35gxOyxjyMLKaX"}},
    {"inTheLast": {"lastPlayed": 4 }}
  ],
  "all": [
    {"isNot": {"rating": 1 }},
    {"isNot": {"artistrating": 1 }},
    {"notContains": {"title": "(Live"}},
    {"notContains": {"filepath": "scores"}},
    {"notContains": {"filepath": "seasonal"}},
    {"notContains": {"discsubtitle": "live"}}
  ],
  "sort": "random"
}
```

This was the rule field before a manual update:
```
navidrome D select name, path, updated_at, rules from playlist where name = 'Radio';
| name  |                  path                  |             updated_at              |                                                  rules                                                  |
|-------|----------------------------------------|-------------------------------------|---------------------------------------------------------------------------------------------------------|
| Radio | /music/playlists/family/Radio.nsp | 2026-07-10 18:12:04.932707865+00:00 | {"any":[{"inPlaylist":{"id":"QtfW3w5x35gxOyxjyMLKaX"}},{"inTheLast":{"lastPlayed":4}}],"sort":"random"} |
```

Update directly:
```
navidrome D update playlist set rules = '{"any":[{"inPlaylist":{"id":"QtfW3w5x35gxOyxjyMLKaX"}},{"inTheLast":{"lastPlayed":4}}],"all":[{"isNot":{"rating":1}},{"isNot":{"artistrating":1}},{"notContains":{"title":"(Live"}},{"notContains":{"filepath":"scores"}},{"contains":{"filepath":"seasonal"}},{"notContains":{"discsubtitle":"live"}}],"sort":"random"}' where name = 'Radio';

| name  |                  path                  |             updated_at              |                                                                                                                                                          rules                                                                                                                                                           |
|-------|----------------------------------------|-------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Radio | /music/playlists/family/Radio.nsp | 2026-07-10 21:40:08.073492935+00:00 | {"any":[{"inPlaylist":{"id":"QtfW3w5x35gxOyxjyMLKaX"}},{"inTheLast":{"lastPlayed":4}}],"all":[{"isNot":{"rating":1}},{"isNot":{"artistrating":1}},{"notContains":{"title":"(Live"}},{"notContains":{"filepath":"scores"}},{"contains":{"filepath":"seasonal"}},{"notContains":{"discsubtitle":"live"}}],"sort":"random"} |
```

And it seems to work fine, I get the results I expect after the update where I didnt get them before the manual update. For example without doing the manual DB update to add the full rule I would get songs where the discsubtitle contains live. And I dont after the update.

### Expected Behavior

I think the main question is if the scanner is doing some parsing to only load one rule group from playlists that contain any + all? And should it be, given that one of the documented examples contains any + all, and it works with a manual DB update to add the full rule?

Or have I missed something entirely?

### Steps To Reproduce

1. Use the "Overplayed Favorites" example from the Smart Playlists documentation page
2. Check rules column from playlists table after its loaded
3. Rule does not contain "any"
4. Insert full rule
5. Seems to work as expected

### Environment

```markdown
- OS: Fedora CoreOS with Podman 5.x
- Browser: Firefox
- Client: Symfonium v15 beta 10, direct DB queries
```

### How Navidrome is installed?

Docker

### Configuration

```toml
Note that this is running in a Podman Quadlet

[Unit]
Description=navidrome container
After=local-fs.target
After=mount-podman-secure.service
After=remote-cryptsetup.target

# Forces systemd to keep trying to start service forever, each RestartSec seconds
StartLimitIntervalSec=0

[Container]
Image=docker.io/deluan/navidrome:latest
ContainerName=navidrome
AutoUpdate=registry
LogDriver=journald
Label=app=navidrome

Environment=TZ=America/Vancouver
Environment=ND_PORT=4533
Environment=ND_ENABLEINSIGHTSCOLLECTOR=true
Environment=ND_LOGFILE=/logs/navidrome.log
Environment=ND_LOGLEVEL="debug"
Environment=ND_BASEURL="https://redacted"
Environment=ND_AUTHREQUESTLIMIT=5
Environment=ND_AUTHWINDOWLENGTH="20s"
Environment=ND_AUTOTRANSCODEDOWNLOAD=true
Environment=ND_MUSICFOLDER="/music/playlists"
Environment=ND_SCANNER_SCANONSTARTUP=true
Environment=ND_PLAYLISTSPATH="admin:me:family"
Environment=ND_AUTOIMPORTPLAYLISTS=true
Environment=ND_ENABLEARTWORKPRECACHE=true
Environment=ND_ENABLEREPLAYGAIN=true
Environment=ND_IMAGECACHESIZE="1024MB"
Environment=ND_BACKUP_PATH="/backups"
Environment=ND_BACKUP_SCHEDULE="30 23 * * *"
Environment=ND_BACKUP_COUNT="2"
Environment=ND_LASTFM_ENABLED="true"
Environment=ND_LASTFM_APIKEY=redacted
Environment=ND_PLUGINS_ENABLED=true
Environment=ND_PLUGINS_AUTORELOAD=true
# Needed for audiomuseai-nv
Environment=ND_DEVARTISTINFOTIMETOLIVE=1s
Environment=ND_AGENTS=audiomuseai,lastfm,listenbrainz


Secret=navidrome-lastfm-api-secret,type=env,target=ND_LASTFM_SECRET

Network=pasta
PublishPort=4533:4533

UserNS=auto:size=65536,uidmapping=0:@1010:1,gidmapping=0:@1010:1

Volume=navidrome-data:/data:Z,U
Volume=/var/mnt/music:/music:ro
Volume=navidrome-logs:/logs:z,U
Volume=navidrome-backups:/backups:z,U

[Service]
##Type=oneshot  #Use for go-acme style services
##RemainAfterExit=yes  #Use for go-acme style services
TimeoutStartSec=120
Restart=on-failure
RestartSec=10
RestartSteps=10
RestartMaxDelaySec=120
ExecStartPre=/usr/bin/systemctl is-active mount-podman-secrets.service
[Install]
WantedBy=multi-user.target default.target


--------------------------------------------

# Navidrome Configuration
# Generated on 2026-07-10T22:01:36.423Z

Address = "0.0.0.0"
Agents = "audiomuseai,lastfm,listenbrainz"
AlbumPlayCountMode = "absolute"
ArtistArtPriority = "artist.*, album/artist.*, external"
ArtistImageFolder = ""
AuthRequestLimit = 5
AuthWindowLength = 20000000000
AutoImportPlaylists = true
AutoTranscodeDownload = true
BaseHost = "redacted"
BasePath = ""
BaseScheme = "https"
BaseURL = "https://redacted"
CacheFolder = "/data/cache"
CoverArtPriority = "cover.*, folder.*, front.*, embedded, external"
CoverArtQuality = 75
DataFolder = "/data"
DbPath = "/data/navidrome.db?cache=shared&_busy_timeout=15000&_journal_mode=WAL&_foreign_keys=on&synchronous=normal"
DefaultDownloadableShare = false
DefaultDownsamplingFormat = "opus"
DefaultLanguage = ""
DefaultPlaylistPublicVisibility = false
DefaultShareExpiration = 31536000000000000
DefaultTheme = "Dark"
DefaultUIVolume = 100
DiscArtPriority = "disc*.*, cd*.*, cover.*, folder.*, front.*, discsubtitle, embedded"
EnableArtworkPrecache = true
EnableArtworkUpload = true
EnableCoverAnimation = true
EnableDownloads = true
EnableExternalServices = true
EnableFavourites = true
EnableGravatar = false
EnableInsightsCollector = true
EnableLogRedacting = true
EnableM3UExternalAlbumArt = false
EnableMediaFileCoverArt = true
EnableNowPlaying = true
EnableReplayGain = true
EnableScrobbleHistory = true
EnableSharing = true
EnableStarRating = true
EnableTranscodingConfig = false
EnableUserEditing = true
EnableWebPEncoding = false
EnforceNonRootUser = false
FFmpegPath = ""
GATrackingID = ""
IgnoredArticles = "The El La Los Las Le Les Os As O A"
ImageCacheSize = "1024MB"
IndexGroups = "A B C D E F G H I J K L M N O P Q R S T U V W X-Z(XYZ) [Unknown]([)"
LogFile = "/logs/navidrome.log"
LogLevel = "debug"
LyricsPriority = ".ttml,.yaml,.yml,.elrc,.lrc,.srt,.txt,embedded"
MaxImageUploadSize = "10MB"
MaxSidebarPlaylists = 100
MPVCmdTemplate = "mpv --audio-device=%d --no-audio-display %f --input-ipc-server=%s"
MPVPath = ""
MusicFolder = "/music/playlists"
PasswordEncryptionKey = ""
PlaylistsPath = "admin:me:family"
Port = 4533
PreferSortTags = false
RecentlyAddedByModTime = false
SessionTimeout = 172800000000000
ShareURL = ""
SmartPlaylistRefreshDelay = 5000000000
TLSCert = ""
TLSKey = ""
TranscodingCacheSize = "100MB"
UICoverArtSize = 300
UILoginBackgroundURL = "/backgrounds"
UIPlaybackReportInterval = 60000000000
UISearchDebounceMs = 200
UIWelcomeMessage = ""
UnixSocketPerm = 0660

# Development Flags (subject to change/removal)
# These are experimental settings and may be removed in future versions

DevActivityPanel = true
DevActivityPanelUpdateRate = 300000000
DevAlbumInfoTimeToLive = 604800000000000
DevArtistInfoTimeToLive = 1000000000
DevArtworkMaxRequests = 2
DevArtworkThrottleBacklogLimit = 100
DevArtworkThrottleBacklogTimeout = 60000000000
DevArtworkThrottleBuffered = true
DevAutoCreateAdminPassword = ""
DevAutoLoginUsername = ""
DevEnableMediaFileProbe = true
DevEnablePlayerInsights = true
DevEnablePluginsInsights = true
DevEnableProfiler = false
DevExternalArtistFetchMultiplier = 1.5
DevExternalScanner = true
DevInsightsInitialDelay = 1800000000000
DevLogSourceLine = false
DevNewEventStream = true
DevOffsetOptimize = 50000
DevOptimizeDB = true
DevPluginCompilationTimeout = 60000000000
DevPreserveUnicodeInExternalCalls = false
DevScannerThreads = 5
DevSelectiveWatcher = true
DevShowArtistPage = true
DevSidebarPlaylists = true
DevUIShowConfig = true

[Backup]
Count = 2
Path = "/backups"
Schedule = "30 23 * * *"

[Deezer]
Enabled = true
Language = "en"
Languages = [ "en" ]

[ExtAuth]
LogoutURL = ""
TrustedSources = ""
UserHeader = "Remote-User"

[HTTPHeaders]
FrameOptions = "DENY"

[Inspect]
BacklogLimit = 100
BacklogTimeout = 60000000000
Enabled = true
MaxRequests = 1

[Jukebox]
AdminOnly = true
Default = ""
Devices = [ ]
Enabled = false

[LastFM]
ApiKey = "c******************************e"
Enabled = true
Language = "en"
Languages = [ "en" ]
ScrobbleFirstArtistOnly = false
Secret = "7******************************2"

[ListenBrainz]
ArtistAlgorithm = "session_based_days_9000_session_300_contribution_5_threshold_15_limit_50_skip_30"
BaseURL = "https://api.listenbrainz.org/1/"
Enabled = true
TrackAlgorithm = "session_based_days_9000_session_300_contribution_5_threshold_15_limit_50_skip_30"

[Matcher]
FuzzyThreshold = 85
PreferStarred = true

[PID]
Album = "musicbrainz_albumid|albumartistid,album,albumversion,releasedate"
Track = "musicbrainz_trackid|albumid,discnumber,tracknumber,title"

[Plugins]
AutoReload = true
CacheSize = "200MB"
Enabled = true
Folder = "/data/plugins"
LogLevel = ""

[Prometheus]
Enabled = false
MetricsPath = "/******s"
Password = ""

[Scanner]
ArtistJoiner = " • "
ArtistSplitExceptions = [ ]
Enabled = true
Extractor = "taglib"
FollowSymlinks = true
GenreSeparators = ""
GroupAlbumReleases = false
IgnoreDotFolders = true
PurgeMissing = "never"
ScanOnStartup = true
Schedule = ""
WatcherWait = 5000000000

[Search]
Backend = "fts"
FullString = false

[Subsonic]
AppendAlbumVersion = true
AppendSubtitle = true
ArtistParticipations = false
DefaultReportRealPath = false
EnableAverageRating = true
LegacyClients = "DSub"
MinimalClients = "SubMusic"

[Tags]
"missingtracks.Aliases" = [ "missingtracks" ]
"missingtracks.MaxLength" = 50
```

### Relevant log output

```shell

```

### Anything else?

Thank you for Navidrome, it is wonderful!

### Code of Conduct

- [x] I agree to follow Navidrome's Code of Conduct

### PR DESCRIPTION
**Pull request by deluan**
### Description

A Smart Playlist (`.nsp`) that declared **both** a top-level `any` and a top-level `all` group was imported by silently keeping only the `any` group and discarding `all` (regardless of key order in the file). Because `criteria.Criteria` stores a single top-level `Expression`, it cannot represent two top-level groups, and `Criteria.UnmarshalJSON` picked `any` without reporting the dropped rules — resulting in silent data loss during the scan.

This PR makes `Criteria.UnmarshalJSON` return an error when both `all` and `any` are present at the top level. The scanner now fails loudly (logging the playlist as invalid, exactly like any other malformed `.nsp`) instead of quietly losing rules. Since the validation lives at the single `Criteria` unmarshalling choke point, it also covers the native-API playlist-rules path, not just file import.

The correct way to combine the two logics is to nest one group inside the other, as shown in the documented examples (e.g. "80's Top Songs"). A companion documentation fix corrects the misleading example that triggered this report.

### Related Issues

Fixes #5757

### Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Documentation update
- [ ] Refactor
- [ ] Other (please describe):

### Checklist

- [x] My code follows the project's coding style
- [x] I have tested the changes locally
- [ ] I have added or updated documentation as needed
- [x] I have added tests that prove my fix/feature works (or explain why not)
- [x] All existing and new tests pass

> Documentation is handled in a separate PR against the `navidrome/website` repo (the `.nsp` docs live there).

### How to Test

1. Create a file `Bad.nsp` in a library folder with both groups at the top level:
   ```json
   {
     "name": "Overplayed Favorites",
     "any": [{ "inPlaylist": { "path": "most-played-favorites.nsp" } }],
     "all": [{ "notInPlaylist": { "path": "favorites-not-played-in-4-yrs.nsp" } }],
     "sort": "playCount, lastPlayed"
   }
   ```
2. Trigger a scan. The playlist is **not** imported, and the log shows an `Error parsing playlist` entry mentioning that `all` and `any` cannot both be used at the top level (previously it imported silently with only the `any` rule).
3. Rewrite it with a single top-level `all` (nesting where needed) and rescan — it imports correctly with all rules:
   ```json
   {
     "name": "Overplayed Favorites",
     "all": [
       { "inPlaylist": { "path": "most-played-favorites.nsp" } },
       { "notInPlaylist": { "path": "favorites-not-played-in-4-yrs.nsp" } }
     ],
     "sort": "playCount, lastPlayed"
   }
   ```

### Additional Notes

This changes behavior for any existing `.nsp` that (perhaps unknowingly) relied on the old silent-drop behavior: such a playlist will now fail to import until it is rewritten to use a single top-level group. This is intentional — surfacing the problem is preferable to silently discarding rules. The alternative (auto-combining the two groups) would change matching semantics and require round-trip marshalling support, so the loud-rejection approach was chosen per the discussion on the issue.

### CODE REVIEW

File: model/criteria/criteria.go [hunk 1]
```diff
@@ -114,6 +114,9 @@ func (c *Criteria) UnmarshalJSON(data []byte) error {
 	if err := json.Unmarshal(data, &aux); err != nil {
 		return err
 	}
+	if len(aux.Any) > 0 && len(aux.All) > 0 {
+		return errors.New("invalid criteria json: 'all' and 'any' cannot both be used at the top level; nest one inside the other instead")
+	}
```
[the remark this answers is not in the document]
Comment (deluan) [reply model/criteria/criteria.go]: Good catch — fixed in ef47858f. The guard now decodes `all`/`any` as `json.RawMessage` and rejects when **both keys are present**, rather than checking decoded slice length. So `{"any":[],"all":[...]}` (and a `null` group) are now rejected instead of silently using the other group.

File: model/criteria/criteria_test.go [hunk 2]
```diff
@@ -80,6 +80,16 @@ var _ = Describe("Criteria", func() {
...
 	})
 
+	Context("with both top-level 'all' and 'any'", func() {
+		It("returns an error instead of silently dropping one of the groups", func() {
+			jsonStr := `{"any":[{"inPlaylist":{"path":"a.nsp"}}],"all":[{"notInPlaylist":{"path":"b.nsp"}}]}`
+			var c Criteria
+			err := json.Unmarshal([]byte(jsonStr), &c)
+			gomega.Expect(err).To(gomega.HaveOccurred())
+			gomega.Expect(err.Error()).To(gomega.And(gomega.ContainSubstring("all"), gomega.ContainSubstring("any")))
+		})
```
[the remark this answers is not in the document]
Comment (deluan) [reply model/criteria/criteria_test.go:90]: Added in ef47858f — a table-driven test now covers the empty/`null` edge cases alongside a populated group: empty `any`, empty `all`, and `null any`, each asserting the mixed-roots error.

File: model/criteria/criteria.go [hunk 3]
```diff
@@ -114,6 +114,9 @@ func (c *Criteria) UnmarshalJSON(data []byte) error {
 	if err := json.Unmarshal(data, &aux); err != nil {
 		return err
 	}
+	if len(aux.Any) > 0 && len(aux.All) > 0 {
```
[the remark this answers is not in the document]
Comment (deluan) [reply model/criteria/criteria.go]: Fixed in ef47858f — switched from slice-length checks to `json.RawMessage` key-presence detection, so a present-but-empty or `null` group (e.g. `{"any":[],"all":[...]}`) is now rejected rather than silently importing only the other group. Added table-driven tests for the empty and `null` cases.
